---
title: "HMMER & UPROC domain analysis"
author: "Gerry Tonkin-Hill"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 12
    fig_height: 8
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.width=12, fig.height=8,
                      echo=TRUE, warning=FALSE, message=FALSE,
                      tidy=TRUE)
```

###Load Libraries
```{r}
library(ggplot2)
library(reshape2)
library(edgeR)
library(plyr)
library(scales)

library(Rsubread)
library(structSSI)
library(data.table)
library(igraph)

library(cluster)
library(fastcluster)
library(RFLPtools)

library(pheatmap)
library(EDASeq)
library(dynamicTreeCut)
library(moduleColor)
library(knitr)

library(DESeq2)


wd = "./"
colors <- c("#0571b0","#ca0020")
```

Here we make use of UPROC to provide an intial filter of the reads before making user of HMMER to allocate the reads more accurately to domain models. The code used to run this pipeline is available from https://github.com/PapenfussLab/reads_to_domains.

```{bash, eval=FALSE}
parallel -j 5 /home/users/allstaff/tonkin-hill.g/transcriptome_manuscript_response/hmmerUproc/run_UPROCHMMERR.sh ::: /stornext/Bioinf/data/bioinf-data/Papenfuss_lab/projects/malaria/duffy_lab/clinical_isolate_RNA_seq_severe_nonsevere/data/nonsevere/I*_R1_*.fastq

parallel -j 5 /home/users/allstaff/tonkin-hill.g/transcriptome_manuscript_response/hmmerUproc/run_UPROCHMMERR.sh ::: /stornext/Bioinf/data/bioinf-data/Papenfuss_lab/projects/malaria/duffy_lab/clinical_isolate_RNA_seq_severe_nonsevere/data/severe/S*_R1_*.fastq
```

The script run_UPROCHMMERR.sh is copied for reference below
```{bash, eval=FALSE}
set -e

read1=$1

read2=${read1/_R1_/_R2_}

echo "running allocate_reads.py ${read1} ${read2}"

python ~/get_domains/reads_to_domains/reads_to_domains/allocate_reads.py -r $read1 -R $read2 -E 1e-2 -o /home/users/allstaff/tonkin-hill.g/transcriptome_manuscript_response/hmmerUproc/
```

##Process resulting counts

First load data into a count matrix
```{r}
files <- Sys.glob("./data/hmmerUproc/*.csv")
domain_count_list <- lapply(files, fread, data.table=FALSE, header=FALSE)
domain_names <- unique(unlist(lapply(domain_count_list, unlist(function(x) x[,1]))))
sample_names <- gsub("_S.*", "", gsub(".*Uproc/", "", files))
unique_sample_names <- unique(sample_names)
domain_counts <- matrix(0, nrow=length(domain_names), ncol=length(unique_sample_names),
                        dimnames = list(domain_names,unique_sample_names))
for (i in 1:length(domain_count_list)){
  domain_counts[domain_count_list[[i]][,1],sample_names[[i]]] <- domain_count_list[[i]][,2]
}

```

Now we can process it using the default DESeq2 pipeline as was done for the aligned Rask domain counts.

Again we remove samples SFD1, SFM9 and SFC.025 as they were not sequenced to a sufficient depth.
```{r}
setwd(wd)
domain_counts <- domain_counts[,!(colnames(domain_counts) %in% c('SFD1-CM', 'SFM9', 'SFC-025'))]
```

```{r}
x_domains <- DGEList(counts=domain_counts)
x_domains <- calcNormFactors(x_domains, method="RLE")
categories <- as.factor(substring(rownames(x_domains$samples),1,1))
```

###A PCA plot of normalised counts
```{r}
plotPCA(cpm(x_domains,normalized.lib.sizes=TRUE, log=T, prior.count=0.5), col=colors[categories], cex=0.8, isLog=TRUE)
title("VAR PCA - Previous Domains")
```

###Investigate differential expression
```{r}
colData <- data.frame(disease=categories)
rownames(colData) <- colnames(x_domains$counts)
dds_dom <- DESeqDataSetFromMatrix(countData = x_domains$counts
                              , colData = colData
                              , design = ~disease)
dds_dom <- estimateSizeFactors(dds_dom)
dds_dom <- DESeq(dds_dom) 
res <- results(dds_dom,contrast=c("disease","S","I"), pAdjustMethod = "BH")
summary(res, alpha=0.05)
```

##A heatmap of the Up-regulated domains
```{r}
res_domains <- res
resOrdered <- res_domains[order(res_domains$padj),]
resSig <- subset(resOrdered, padj < 0.05)

resSigLFC <- data.frame(subset(resSig, log2FoldChange > 0))

select <- order(rowMeans(counts(dds_dom, normalized=TRUE)), decreasing=TRUE)

nt <- normTransform(dds_dom)
log2.norm.counts <- assay(nt)[select,]
log2.norm.counts <- assay(nt)[rownames(nt) %in% rownames(resSigLFC),]
annotData <- colData

disease        <- colors
names(disease) <- c("non-severe", "severe")
anno_colors <- list(phenotype = disease)
annotData$phenotype[annotData$disease=="I"] <- "non-severe" 
annotData$phenotype[annotData$disease=="S"] <- "severe"
# annotData <- annotData[annotData$disease=="S",]
annotData$disease <- NULL
pheatmap(log2.norm.counts, cluster_rows=TRUE, show_rownames=TRUE, cluster_cols=TRUE, annotation_col=annotData, annotation_colors = anno_colors)
```